src/ir/stack-utils.cpp

```diff
@@ -0,0 +1,278 @@
+  {
+    size_t stackSize = 0;
+    size_t produced = 0;
+    Expression* lastUnreachable = nullptr;
+    processBlock([&](Expression* expr, StackSignature sig) {
+      // Consume params
+      if (sig.params.size() > stackSize) {
+        // We need more values than are available, so they must come from the
+        // last unreachable.
+        assert(lastUnreachable);
+        produced += sig.params.size() - stackSize;
+        stackSize = 0;
+      } else {
+        stackSize -= sig.params.size();
+      }
+
+      // Handle unreachable or produce results
+      if (sig.unreachable) {
+        if (lastUnreachable) {
+          producedByUnreachable[lastUnreachable] = produced;
+          produced = 0;
+        }
+        assert(produced == 0);
+        lastUnreachable = expr;
+        stackSize = 0;
+      } else {
+        stackSize += sig.results.size();
+      }
+    });
+
+    // Finish record for final unreachable
+    if (lastUnreachable) {
+      producedByUnreachable[lastUnreachable] = produced;
+    }
+  }
+
+  // Take another pass through the block, recording srcs and dests.
+  std::vector<Location> values;
+  Expression* lastUnreachable = nullptr;
+  processBlock([&](Expression* expr, StackSignature sig) {
+    assert((sig.params.size() <= values.size() || lastUnreachable) &&
+           "Block inputs not yet supported");
+
+    // Unreachable instructions consume all available values
+    size_t consumed = sig.unreachable
+                        ? std::max(values.size(), sig.params.size())
+                        : sig.params.size();
+
+    // We previously calculated how many values unreachable instructions produce
+    size_t produced =
+      sig.unreachable ? producedByUnreachable[expr] : sig.results.size();
+
+    srcs[expr] = std::vector<Location>(consumed);
+    dests[expr] = std::vector<Location>(produced);
+
+    // Consume values
+    assert(sig.params.size() <= consumed);
+    size_t unreachableConsumed =
+      consumed > values.size() ? consumed - values.size() : 0;
+    for (Index i = 0; i < consumed; ++i) {
+      if (i < unreachableConsumed) {
+        // This value comes from the polymorphic stack of the last unreachable
+        // because the stack did not have enough values to satisfy this
+        // instruction.
+        assert(consumed == sig.params.size());
+        assert(lastUnreachable);
+        assert(producedByUnreachable[lastUnreachable] >= unreachableConsumed);
+        Index destIndex =
+          producedByUnreachable[lastUnreachable] - unreachableConsumed + i;
+        Type type = sig.params[i];
+        srcs[expr][i] = {lastUnreachable, destIndex, type, true};
+        dests[lastUnreachable][destIndex] = {expr, i, type, false};
+      } else {
+        // A normal value from the value stack
+        bool unreachable = sig.params.size() + i < consumed;
+        auto& src = values[values.size() + i - consumed];
+        srcs[expr][i] = {src.expr, src.index, src.type, false};
```

Comment:
```suggestion
        bool unreachable = i > sig.param.size();
        auto& src = values[values.size() + i - consumed];
        srcs[expr][i] = src;
```
Can this be correct too? I don't understand 
```
bool unreachable = sig.params.size() + i < consumed;
```
This line very well.

Reply:
Say `sig.params.size() == 3` and `consumed == 5`. The way the code is written, `unreachable` will be true when `i` is 0 or 1 and `false` when it is in [2,5). This is correct because the deeper values are consumed by the current instruction's polymorphic behavior without being used (`unreachable` true) and the shallower values are used directly by the current instruction (`unreachable` false) and we are iterating through consumed values from deepest to shallowest.

I'll tweak this code to try to make it clearer. Also, you're totally right about the potential simplification in the last line.